src/torchjd/sparse/_diagonal_sparse_tensor.py

```diff
@@ -0,0 +1,963 @@
+from torch.ops import aten  # type: ignore
+from torch.utils._pytree import tree_flatten, tree_map, tree_unflatten
+
+
+class DiagonalSparseTensor(Tensor):
+    _HANDLED_FUNCTIONS = dict()
+
+    @staticmethod
+    def __new__(cls, physical: Tensor, v_to_ps: list[list[int]]):
+        # At the moment, this class is not compositional, so we assert
+        # that the tensor we're wrapping is exactly a Tensor
+        assert type(physical) is Tensor
+
+        # Note [Passing requires_grad=true tensors to subclasses]
+        # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+        # Calling _make_subclass directly in an autograd context is
+        # never the right thing to do, as this will detach you from
+        # the autograd graph.  You must create an autograd function
+        # representing the "constructor" (NegativeView, in this case)
+        # and call that instead.  This assert helps prevent direct usage
+        # (which is bad!)
+        assert not physical.requires_grad or not torch.is_grad_enabled()
+
+        shape = [prod(physical.shape[i] for i in dims) for dims in v_to_ps]
+        return Tensor._make_wrapper_subclass(
+            cls, shape, dtype=physical.dtype, device=physical.device
+        )
+
+    def __init__(self, physical: Tensor, v_to_ps: list[list[int]]):
+        """
+        This constructor is made for specifying physical and v_to_ps exactly. It should not modify
+        it.
+
+        For this reason, another constructor will be made to either modify the physical / v_to_ps to
+        simplify the result, or to create a dense tensor directly if it's already dense. It could
+        also be responsible for sorting the first apparition of each physical dim in the flattened
+        v_to_ps.
+        """
+
+        if any(s == 1 for s in physical.shape):
+            raise ValueError(
+                "physical must not contain any dimension of size 1. Found physical.shape="
+                f"{physical.shape}."
+            )
+        if not all(all(0 <= dim < physical.ndim for dim in dims) for dims in v_to_ps):
+            raise ValueError(
+                f"Elements in v_to_ps must map to dimensions in physical. Found {v_to_ps}."
+            )
+        if len(set().union(*[set(dims) for dims in v_to_ps])) != physical.ndim:
+            raise ValueError("Every dimension in physical must appear at least once in v_to_ps.")
+
+        if v_to_ps != encode_v_to_ps(v_to_ps)[0]:
+            raise ValueError(
+                f"v_to_ps elements are not encoded by first appearance. Found {v_to_ps}."
+            )
+
+        if any(len(group) != 1 for group in get_groupings(v_to_ps)):
+            raise ValueError(f"Dimensions must be maximally grouped. Found {v_to_ps}.")
+
+        self.physical = physical
+        self.v_to_ps = v_to_ps
+        pshape = list(self.physical.shape)
+        self.strides = tensor([strides_v2(pdims, pshape) for pdims in self.v_to_ps])
```

Comment:
Could we make `strides_v2` return a `Tensor` directly so that if we change from `v_to_ps` to pure `stride` description, we can then already assume it is an array (we could have a function to build a stride from a `v_to_ps` and `physical.shape`)